deep_quoridor/src/quoridor.py

```diff
@@ -0,0 +1,293 @@
+    PLAYER1_WALL = 10
+    PLAYER2_WALL = 11
+
+    def __init__(self, board_size: int, max_walls: int):
+        self.board_size = board_size
+        self.max_walls = max_walls
+
+        self._grid = np.full((board_size * 2 - 1, board_size * 2 - 1), Board.FREE, dtype=np.int8)
+        self._player_positions = np.array([(0, board_size // 2), (board_size - 1, board_size // 2)])
+        for player, position in enumerate(self._player_positions):
+            self._grid[*position * 2] = player
+        self._walls_remaining = [self.max_walls, self.max_walls]
+
+    def get_player_position(self, player: int) -> Position:
+        """
+        Get the position of the player's pawn.
+        """
+        try:
+            return self._player_positions[player]
+        except IndexError:
+            raise ValueError(f"Invalid player {player}")
+
+    def move_player(self, player: int, new_position: Position):
+        """
+        Move the player's pawn. Doesn't check if the move is valid according to game rules.
+        """
+        new_position = np.asarray(new_position, dtype=np.uint8)
+
+        try:
+            old_position = self._player_positions[player]
+        except IndexError:
+            raise ValueError(f"Invalid player {player}")
+
+        try:
+            self._grid[*old_position * 2] = Board.FREE
+            self._grid[*new_position * 2] = player
+            self._player_positions[player] = new_position
+        except IndexError:
+            return Board.OUT_OF_BOUNDS
+
+    def get_cell(self, position: Position) -> int:
+        """
+        Get the value of a "player cell" in the grid.
+        """
+        position = np.asarray(position, dtype=np.uint8)
+        try:
+            return self._grid[*position * 2]
+        except IndexError:
+            return Board.OUT_OF_BOUNDS
+
+    def add_wall(self, player: int, position: Position, orientation: WallOrientation):
+        """
+        Mark the grid cells corresponding to the wall as occupied.
+        """
+        position = np.asarray(position, dtype=np.uint8)
+
+        if self._walls_remaining[player] < 1:
+            raise RuntimeError("Cannot place wall, no walls remaining for player {player}")
+        try:
+            self._grid[self._get_wall_slices(position, orientation)] = 10 + player
+        except IndexError:
+            return Board.OUT_OF_BOUNDS
+
+        self._walls_remaining[player] -= 1
+
+    def can_place_wall(self, position: Position, orientation: WallOrientation) -> bool:
+        """
+        Returns True if the wall can be placed at the given position and orientation.
+
+        Checks that the player has walls remaining, the wall is within the bounds of the board, and doesn't overlap with other walls.
+        """
+        position = np.asarray(position, dtype=np.uint8)
+        try:
+            return (self._grid[self._get_wall_slices(position, orientation)] == Board.FREE).all()
+        except IndexError:
+            return False
+
```

Comment:
if there's a problem with the indexes, that sounds like a bug, we should let the exception go through or detect it early, so we're not masking the bug

Reply:
What do you mean by "problem with the indexes"? All jumps that can every be valid are in the `_jump_checks` dictionary, so this is checking whether this is a valid move. For example moving 3 spaces is never valid, and this catches that because it isn't in the dictionary.

Reply:
The comment here doesn't seem to match the code, since I don't see any jumps here.  
Anyway, it's safer to explicitely check for the case rather than rely on exceptions.  E.g. in this case, if the position is inside the board (which probably should be checked earlier), then I would expect `_get_wall_slices` to return a valid index in the grid.

Reply:
+1 here on checking the indexes prior to use instead of catching the index error

Reply:
Done